libs/execution/src/lib/execution-context.ts

```diff
@@ -128,6 +129,7 @@ export class ExecutionContext {
     assert(!isPipelineDefinition(currentNode));
     assert(!isExpressionConstraintDefinition(currentNode));
     assert(!isTransformDefinition(currentNode));
+    assert(isReference(currentNode.type));
```

Comment:
Probably can delete all other asserts because this already narrows it down to a reference.